pkg/util/kms.go

```diff
@@ -0,0 +1,277 @@
+		}
+		config[vaultClientCert] = clientCertFileAddr
+
+	}
+	if clientKeySecretName, ok := config[vaultClientKey]; ok {
+		secret.Name = clientKeySecretName.(string)
+		if !KubeCheckOptional(secret) {
+			return fmt.Errorf(`❌ Could not find secret %q in namespace %q`, secret.Name, secret.Namespace)
+		}
+		clientKeyFileAddr, err := writeCrtsToFile(secret.Name, namespace, secret.Data["key"], vaultClientKey)
+		if err != nil {
+			return fmt.Errorf("can not write crt %v to file %v", vaultClientKey, err)
+		}
+		config[vaultClientKey] = clientKeyFileAddr
+	}
+	return nil
+}
+
+// PutSecret writes the secret to the secrets store
+func PutSecret(client secrets.Secrets, secretName, secretValue, secretPath string) error {
+
+	keyContext := map[string]string{}
+	data := make(map[string]interface{})
+	data[secretName] = secretValue
+
+	err := client.PutSecret(secretPath, data, keyContext)
+	if err != nil {
+		log.Errorf("KMS PutSecret: secret path %v value %v, error %v", secretPath, secretValue, err)
+		return err
+	}
+
+	return nil
+}
+
+// GetSecret reads the secret to the secrets store
+func GetSecret(client secrets.Secrets, secretName, secretPath string) (string, error) {
+	keyContext := map[string]string{}
+	s, err := client.GetSecret(secretPath, keyContext)
+	if err != nil {
+		log.Errorf("KMS GetSecret: secret path %v, error %v", secretPath, err)
+		return "", err
+	}
+
+	return s[secretName].(string), nil
+}
+
+// DeleteSecret deletes the secret from the secrets store
+func DeleteSecret(client secrets.Secrets, secretPath string) error {
+	keyContext := map[string]string{}
+	err := client.DeleteSecret(secretPath, keyContext)
+	if err != nil {
+		log.Errorf("KMS DeleteSecret: secret path %v, error %v", secretPath, err)
+		return err
+	}
+	return nil
+}
+
+// BuildExternalSecretPath builds a string that specifies the root key secret path
+func BuildExternalSecretPath(kms nbv1.KeyManagementServiceSpec, uid string) (string) {
+	secretPath := rootSecretPath + "/rootkeyb64-" + uid
+	return secretPath
+}
+
+// isVaultKMS return true if kms provider is vault
+func isVaultKMS(provider string) bool {
+	return provider == KmsProviderVault
+}
+
+// ValidateConnectionDetails return error if kms connection details are faulty
+func ValidateConnectionDetails(config map[string]string, tokenSecretName string, namespace string) error {
+	// validate auth token
+	if tokenSecretName == "" {
+		return fmt.Errorf("kms token is missing")
+	}
+	secret := KubeObject(bundle.File_deploy_internal_secret_empty_yaml).(*corev1.Secret)
+	secret.Namespace = namespace
+	secret.Name = tokenSecretName
+
+	if !KubeCheck(secret) {
+		return fmt.Errorf(`❌ Could not find secret %q in namespace %q`, secret.Name, secret.Namespace)
+	}
+
+	token, ok := secret.StringData["token"]
+	if !ok || token == "" {
+		return fmt.Errorf("kms token in token secret is missing")
+	}
+	// validate connection details
+	providerType := config[KmsProvider]
+	if !isVaultKMS(providerType) {
```

Comment:
this should be moved to validateVaultConnectionDetails

Reply:
Is not it in ValidateConnectionDetails()?

Reply:
Is not it in ValidateConnectionDetails()?